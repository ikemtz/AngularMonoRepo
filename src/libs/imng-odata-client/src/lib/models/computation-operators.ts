export declare enum ComputationOperators {
  Multiply = 'mul',
  Divide = 'div',
  Add = 'add',
  Subtract = 'sub',
  Modulus = 'mod',
}
