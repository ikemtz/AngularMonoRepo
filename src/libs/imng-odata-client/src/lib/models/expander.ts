import { ODataQuery } from './odata-query';

export interface Expander extends ODataQuery {
  table: string;
}

export function isExpander(source: string | Expander): source is Expander {
  return !!(source as Expander)?.table;
}
