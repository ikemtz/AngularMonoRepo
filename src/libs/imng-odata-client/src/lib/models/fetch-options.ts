export interface FetchOptions {
  /**
   * Collection of child table properties that are rendered on a grid
   */
  boundChildTableProperties?: BoundChildTableProperty[];
  /**
   * Collection of property names that are of type: nullable Date
   */
  dateNullableProps?: string[];
  /**
   * Collection of property names that are of type: nullable UTC DateTime
   *    Note: these properties will be converted to local date time
   */
  utcNullableProps?: string[];
  /**
   * Set to true to force request via cache-busting.
   * Default: false;
   */
  bustCache?: boolean;
}

export interface BoundChildTableProperty {
  /**
   * This is the child table name
   */
  table: string;
  /**
   * This is the field name on the child table that is bound
   */
  field: string;
  /**
   * This controls if 'all' child records have to meet the the conditions
   * or if 'any' of the child records have to meet the condition.
   * Default: 'any'
   */
  linqOperation?: 'all' | 'any';
}
