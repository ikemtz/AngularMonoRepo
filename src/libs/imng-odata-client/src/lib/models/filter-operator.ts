import { IdType } from 'imng-nrsrx-client-utils';

export interface IFilterOperator {
  name: string;
  toODataString:
    | ((field: string, value?: IdType) => string)
    | ((field: string, value?: IdType[]) => string);
}
