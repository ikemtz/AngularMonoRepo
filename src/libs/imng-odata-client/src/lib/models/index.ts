export * from './child-filter';
export * from './composite-filter';
export * from './computation-operators';
export * from './computation';
export * from './expander';
export * from './fetch-options';
export * from './filter-operators';
export * from './filter';
export * from './odata-query';
export * from './odata-result';
export * from './sort';
export * from './filter-serializers';
