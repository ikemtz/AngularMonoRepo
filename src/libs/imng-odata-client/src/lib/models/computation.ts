import { ComputationOperators } from './computation-operators';

export interface Computation {
  /** This is the first value to be used in the computation */
  fieldA: string | number;
  /** This is the second value to be used in the computation */
  fieldB: string | number;
  operator: ComputationOperators | string; //NOSONAR

  /** This MUST differ from the names of declared or dynamic properties of the identified resources. */
  alias: string;
}

export function isComputation(
  source: string | Computation,
): source is Computation {
  return !!(source as Computation)?.operator;
}
