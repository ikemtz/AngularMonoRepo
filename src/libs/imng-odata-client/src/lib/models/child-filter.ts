import { CompositeFilter } from './composite-filter';
import { Filter } from './filter';

export interface ChildFilter extends Filter {
  /**
   * Used to support child entity filtering:
   * Child table name/navigation property
   * */
  childTable: string;
  /**
   * Used to support child entity filtering:
   * Specifies whether all or some of the child records have to match the given criteria
   * */
  linqOperation: 'all' | 'any';
}

export function isChildFilter(
  source: CompositeFilter | Filter | ChildFilter,
): source is ChildFilter {
  return !!(source as ChildFilter)?.childTable;
}
