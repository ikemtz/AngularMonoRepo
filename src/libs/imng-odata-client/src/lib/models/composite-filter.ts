import { ChildFilter } from './child-filter';
import { Filter } from './filter';

/**
 * A complex filter expression.
 */
export interface CompositeFilter {
  /**
   * The logical operation to use when the `filter.filters` option is set.
   *
   * The supported values are:
   * * `"and"`
   * * `"or"`
   */
  logic: 'or' | 'and';
  /**
   * The nested filter expressions
   */
  filters: Array<Filter | ChildFilter | CompositeFilter>;
}

export function isCompositeFilter(
  source: CompositeFilter | Filter | ChildFilter,
): source is CompositeFilter {
  return !!(source as CompositeFilter)?.filters;
}
