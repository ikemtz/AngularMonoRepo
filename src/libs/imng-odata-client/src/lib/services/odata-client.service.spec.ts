import { HttpClient } from '@angular/common/http';
import { TestBed } from '@angular/core/testing';
import { readFirst } from 'imng-ngrx-utils/testing';
import { of } from 'rxjs';
import { FilterOperators, ODataResult } from '../models';

import { ODataClientService } from './odata-client.service';

const testODataResult = (): ODataResult<{
  id: string;
  dt?: Date;
  dtDate?: Date;
  dtUtc?: Date;
}> => ({
  count: 2,
  value: [
    { id: 'x', dt: new Date(), dtDate: new Date(), dtUtc: new Date() },
    { id: 'y' },
  ],
});

describe('ODataClientService', () => {
  let service: ODataClientService;
  let httpClient: HttpClient;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [{ provide: HttpClient, useValue: {} }],
    });
    service = TestBed.inject(ODataClientService);
    httpClient = TestBed.inject(HttpClient);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  it('should serialize ODataQueries with ODataResults', () => {
    let queryString = '';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (httpClient as any).get = jest.fn((x) => {
      queryString = x;
      return of(testODataResult());
    });
    readFirst(
      service.fetch(
        'imng.com',
        {
          select: ['A', 'b', '890'],
          top: 123,
          skip: 456,
        },
        {
          dateNullableProps: ['dtDate', 'dt'],
          utcNullableProps: ['dtUtc', 'dt'],
        },
      ),
    );
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with array results', () => {
    let queryString = '';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (httpClient as any).get = jest.fn((x) => {
      queryString = x;
      return of(testODataResult().value);
    });
    readFirst(
      service.fetch('imng.com', {
        select: ['A', 'b', '890'],
        top: 123,
        skip: 456,
      }),
    );
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with null results', () => {
    let queryString = '';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (httpClient as any).get = jest.fn((x) => {
      queryString = x;
      return of(null);
    });
    readFirst(
      service.fetch('imng.com', {
        select: ['A', 'b', '890'],
        top: 123,
        skip: 456,
        count: false,
      }),
    );
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with string Filters', () => {
    const queryString = service.getODataString({
      select: ['A', 'b', '890'],
      filter: {
        logic: 'and',
        filters: [
          { field: 'A', operator: FilterOperators.equals, value: 1 },
          { field: 'b', operator: FilterOperators.greaterThan, value: '2' },
          { field: 'b', operator: FilterOperators.notNull, value: '2' },
          {
            field: 'b',
            operator: FilterOperators.greaterThanOrEquals,
            value: '2',
          },
          { field: 'b', operator: FilterOperators.in, value: ['2', 3] },
          { field: 'b', operator: FilterOperators.lessThan, value: '2' },
          {
            field: 'b',
            operator: FilterOperators.lessThanOrEquals,
            value: '2',
          },
          {
            logic: 'or',
            filters: [
              {
                field: 'subX',
                operator: FilterOperators.startsWith,
                value: 'b',
              },
              { field: 'subY', operator: FilterOperators.isNull },
              {
                logic: 'or',
                filters: [
                  {
                    childTable: 'sub-z',
                    linqOperation: 'all',
                    field: 'sub-sub-Z',
                    operator: FilterOperators.isNull,
                  },
                ],
              },
            ],
          },
        ],
      },
      top: 123,
      skip: 456,
      expand: [{ table: 'xyz', select: ['id', 'abc'] }],
    });
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize Expanders', () => {
    const queryString = service.getODataString({
      expand: [
        {
          table: 'tableName',
          select: ['A', 'b', '890'],
          expand: [
            {
              table: 'xyz',
              select: ['id', 'abc'],

              filter: {
                logic: 'and',
                filters: [
                  { field: 'A', operator: FilterOperators.equals, value: 1 },
                  {
                    field: 'b',
                    operator: FilterOperators.greaterThan,
                    value: '2',
                  },
                  { field: 'b', operator: FilterOperators.notNull, value: '2' },
                  {
                    field: 'b',
                    operator: FilterOperators.greaterThanOrEquals,
                    value: '2',
                  },
                  { field: 'b', operator: FilterOperators.in, value: ['2', 3] },
                  {
                    field: 'b',
                    operator: FilterOperators.lessThan,
                    value: '2',
                  },
                  {
                    field: 'b',
                    operator: FilterOperators.lessThanOrEquals,
                    value: '2',
                  },
                  {
                    logic: 'or',
                    filters: [
                      {
                        field: 'subX',
                        operator: FilterOperators.startsWith,
                        value: 'b',
                      },
                      { field: 'subY', operator: FilterOperators.isNull },
                      {
                        logic: 'or',
                        filters: [
                          {
                            childTable: 'sub-z',
                            linqOperation: 'all',
                            field: 'sub-sub-Z',
                            operator: FilterOperators.isNull,
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            },
          ],
        },
      ],
      top: 123,
      skip: 456,
    });
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with a guid Filter', () => {
    const queryString = service.getODataString({
      filter: {
        logic: 'and',
        filters: [
          {
            field: 'xyz',
            operator: FilterOperators.equals,
            value: '12345678-1234-1234-1234-1234567890ab',
          },
        ],
      },
      top: 123,
      skip: 456,
      expand: [{ table: 'xyz', select: ['id', 'abc'] }],
      orderBy: [
        { field: 'xyz', dir: 'desc' },
        { field: 'id', dir: 'asc' },
      ],
    });
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with a date Filter', () => {
    const queryString = service.getODataString({
      select: ['A', 'b', '890'],
      filter: {
        logic: 'and',
        filters: [
          {
            field: 'xyz',
            operator: FilterOperators.equals,
            value: new Date(2022, 2, 2),
          },
        ],
      },
      top: 123,
      skip: 456,
      expand: [{ table: 'xyz', select: ['id', 'abc'] }],
      orderBy: [
        { field: 'xyz', dir: 'desc' },
        { field: 'id', dir: 'asc' },
      ],
    });
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with an empty Filter', () => {
    const queryString = service.getODataString({
      select: ['A', 'b', '890'],
      filter: {
        logic: 'and',
        filters: [],
      },
      top: 123,
      skip: 456,
      expand: [{ table: 'xyz', select: ['id', 'abc'] }],
      orderBy: [
        { field: 'xyz', dir: 'desc' },
        { field: 'id', dir: 'asc' },
      ],
    });
    expect(queryString).not.toContain('?&');
    expect(queryString).not.toContain('timestamp');
    expect(queryString).toMatchSnapshot();
  });

  it('should serialize ODataQueries with cacheBusting', () => {
    const queryString = service.getODataString(
      {
        select: ['A', 'b', '890'],
      },
      { bustCache: true },
    );
    expect(queryString).not.toContain('?&');
    expect(queryString).toContain('timestamp');
    expect(queryString).toHaveLength(55);
  });
});
