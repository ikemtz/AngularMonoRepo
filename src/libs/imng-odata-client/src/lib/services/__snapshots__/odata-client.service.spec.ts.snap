// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ODataClientService should serialize Expanders 1`] = `"$expand=tableName($select=A,b,890;$expand=xyz($filter=(A+eq+1+and+b+gt+'2'+and+b+ne+null+and+b+ge+'2'+and+b+in+('2',3)+and+b+lt+'2'+and+b+le+'2'+and+(startswith(subX,'b')+or+subY+eq+null+or+(sub-z/all(o: o/sub-sub-Z+eq+null))));$select=id,abc))&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries simple expander 1`] = `"$expand=xyz&$count=true"`;

exports[`ODataClientService should serialize ODataQueries simple nested expander 1`] = `"$expand=xyz($expand=abc)&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with ODataResults 1`] = `"imng.com?$select=A,b,890&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with a contains Filter 1`] = `"$filter=((sub-z/all(o: contains(o/x,'z')))+and+(sub-z/all(o: contains(o/y,'z'))))&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with a date Filter 1`] = `"$filter=(xyz+eq+2022-03-02)&$orderby=xyz+desc,id&$select=A,b,890&$expand=xyz($select=id,abc)&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with a guid Filter 1`] = `"$filter=(xyz+eq+12345678-1234-1234-1234-1234567890ab)&$orderby=xyz+desc,id&$expand=xyz($select=id,abc)&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with an empty Filter 1`] = `"$orderby=xyz+desc,id&$select=A,b,890&$expand=xyz($select=id,abc)&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with array results 1`] = `"imng.com?$select=A,b,890&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with expander counts 1`] = `"$expand=xyz($orderby=x+desc;$select=id,abc;$top=1;$count=true)&$count=true"`;

exports[`ODataClientService should serialize ODataQueries with null results 1`] = `"imng.com?$select=A,b,890&$top=123&$skip=456"`;

exports[`ODataClientService should serialize ODataQueries with string Filters 1`] = `"$filter=(A+eq+1+and+b+gt+'2'+and+b+ne+null+and+b+ge+'2'+and+b+in+('2',3)+and+b+lt+'2'+and+b+le+'2'+and+(startswith(subX,'b')+or+subY+eq+null+or+(sub-z/all(o: o/sub-sub-Z+eq+null))))&$select=A,b,890&$expand=xyz($select=id,abc)&$top=123&$skip=456&$count=true"`;

exports[`ODataClientService should serialize ODataQueries without expander counts 1`] = `"$expand=xyz($orderby=x+desc;$select=id,abc;$top=1)&$count=true"`;
